### PR TITLE
8320239: add dynamic switch for JvmtiVTMSTransitionDisabler sync protocol

### PR DESCRIPTION
**Pull request by sspitsyn**
This is an update for a performance/scalability enhancement.

The `JvmtiVTMSTransitionDisabler`sync protocol is on a performance critical path of the virtual threads mount state transitions (VTMS transitions). It has a noticeable performance overhead (about 10%) which contributes to the combined JVMTI performance overhead when Java apps are executed with loaded JVMTI agents.

Please, also see another/related performance issue which contributes around 70% of total performance overhead:
[8308614](https://bugs.openjdk.org/browse/JDK-8308614): Enabling JVMTI ClassLoad event slows down vthread creation by factor 10

Testing:
 - Ran mach5 tiers 1-6 with no regressions noticed.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8320239](https://bugs.openjdk.org/browse/JDK-8320239): add dynamic switch for JvmtiVTMSTransitionDisabler sync protocol (**Enhancement** - P4)

### Reviewers
 * [Leonid Mesnik](https://openjdk.org/census#lmesnik) (@lmesnik - **Reviewer**) ⚠️ Review applies to [a81218fd](https://git.openjdk.org/jdk/pull/16688/files/a81218fdc4ba4a47904fbefe069e74a173e5ac0f)
 * [Patricio Chilano Mateo](https://openjdk.org/census#pchilanomate) (@pchilano - **Reviewer**)
 * [Alex Menkov](https://openjdk.org/census#amenkov) (@alexmenkov - **Reviewer**) ⚠️ Review applies to [a81218fd](https://git.openjdk.org/jdk/pull/16688/files/a81218fdc4ba4a47904fbefe069e74a173e5ac0f)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16688/head:pull/16688` \
`$ git checkout pull/16688`

Update a local copy of the PR: \
`$ git checkout pull/16688` \
`$ git pull https://git.openjdk.org/jdk.git pull/16688/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16688`

View PR using the GUI difftool: \
`$ git pr show -t 16688`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16688.diff">https://git.openjdk.org/jdk/pull/16688.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16688#issuecomment-1814365587)

### CODE REVIEW

File: src/hotspot/share/prims/jvmtiThreadState.cpp [hunk 1]
```diff
@@ -404,8 +422,17 @@ JvmtiVTMSTransitionDisabler::VTMS_transition_enable_for_all() {
 void
 JvmtiVTMSTransitionDisabler::start_VTMS_transition(jthread vthread, bool is_mount) {
   JavaThread* thread = JavaThread::current();
+  oop vt = JNIHandles::resolve_external_guard(vthread);
+
+  if (!sync_protocol_enabled()) {
+    assert(!thread->is_in_VTMS_transition(), "VTMS_transition sanity check");
+    thread->set_is_in_VTMS_transition(true);
+     java_lang_Thread::set_is_in_VTMS_transition(vt, true);
```
Comment (lmesnik) [inline src/hotspot/share/prims/jvmtiThreadState.cpp:430]: indentation is incorrect.

File: src/hotspot/share/prims/jvmtiThreadState.hpp [hunk 2]
```diff
@@ -83,6 +83,8 @@ class JvmtiVTMSTransitionDisabler {
   static volatile int _VTMS_transition_disable_for_all_count; // transitions for all virtual threads are disabled while it is positive
   static volatile bool _SR_mode;                         // there is an active suspender or resumer
   static volatile int _VTMS_transition_count;            // current number of VTMS transitions
+  static int  _sync_protocol_enabled_count;              // current number of JvmtiVTMSTransitionDisablers enabled sync protocol
```
Comment (lmesnik) [inline src/hotspot/share/prims/jvmtiThreadState.hpp:86]: The _sync_protocol_enabled_count and  _sync_protocol_enabled_permanently are read/updated in different threads. How access to them is protected from racing? Might be make sense to add this info in comment?

File: src/hotspot/share/prims/jvmtiThreadState.cpp (hunk 1 above)
Comment (sspitsyn) [reply src/hotspot/share/prims/jvmtiThreadState.cpp:430]: Thank you. Fixed now.

File: src/hotspot/share/prims/jvmtiThreadState.hpp (hunk 2 above)
Comment (sspitsyn) [reply src/hotspot/share/prims/jvmtiThreadState.hpp:86]: Good catch, thanks. My initial intention was to make them volatile with Atomic load/store/update. Fixed now.

**Review by lmesnik — APPROVED** [review]

File: src/hotspot/share/prims/jvmtiThreadState.cpp [hunk 3]
```diff
@@ -404,8 +422,17 @@ JvmtiVTMSTransitionDisabler::VTMS_transition_enable_for_all() {
 void
 JvmtiVTMSTransitionDisabler::start_VTMS_transition(jthread vthread, bool is_mount) {
   JavaThread* thread = JavaThread::current();
+  oop vt = JNIHandles::resolve_external_guard(vthread);
+
+  if (!sync_protocol_enabled()) {
```
Comment (pchilano) [inline src/hotspot/share/prims/jvmtiThreadState.cpp:427]: Isn't the check racy? So we could see that there is no disabler, but before setting anything a disabler can run and succeed. Then we would never notice it and will just return after setting these fields.
I think you need to check this after setting _VTMS_transition_count and the vthread transition bit.

**Review by pchilano — COMMENTED** [review]
Hi Serguei, 

One comment below.

Thanks

**Review by alexmenkov — APPROVED** [review]

File: src/hotspot/share/prims/jvmtiThreadState.cpp (hunk 3 above)
Comment (sspitsyn) [reply src/hotspot/share/prims/jvmtiThreadState.cpp:433]: Thank you for looking at this, Patricio!
Nice catch - fixed now.

**Comment by sspitsyn** [comment]
Leonid, Alex and Patricio thank you for review!

**Pull request closed without merging**